teps/0028-task-results-without-results.md

```diff
@@ -0,0 +1,358 @@
+```
+
+Today, `pipeline` is declared `failure` and stops executing further after task result resolution fails because of
+missing task results. There are many reasons for a missing task result:
+ 
+* a `task` producing task result failed, no result available
+* a `task` producing result was skipped/disabled and no result generated
+* a `task` producing result did not generate that result even without any failure
+
+As we are expanding task results consumption within `finally` tasks and adding support for `continueAfterSkip`,
+it becomes crucial to define how a `pipeline` can handle such missing results. In this proposal, we are evaluating
+different possible solutions and determine which one fits best to address this issue.
+
+## Motivation
+
+Missing task results should not be fatal. Continue executing `pipeline` when a task result is missing.
+
+### Goals
+
+Identify a set of use cases for `finally` and `continueAfterSkip` without task results and design a common solution
+to address those use cases.
+
+### Non-Goals
+
+This proposal is not offering a solution without any use case.
+
+## Proposal
+
+Enable `pipeline` author to specify a default value for a `task result` in case of a missing result. Pipeline author
+can specify default for the `param` referring to task result if that `param` is declared optional in the `task`.
+
+Generally, task author designs a task with set of inputs in the form of `params` while producing some results in the
+form of `results`. When a `task` is incorporated in a `pipeline`, `pipeline` author has control over these `params`
+and can design it so that the `params` are mapped to `pipeline` params or initialized with the `results` of other
+tasks in a `pipeline`. `params` in a `task` can be declared as required and optional. Required `params` does not allow
+any defaults and have to be initialized at the `pipeline` or `pipelineRun` level.
+
+The proposal is designed on the similar lines, when a `param` is declared `optional` at the `task` level and referring to the
+`task result` of any other task in a `pipeline`, is allowed to have a default at the `pipelineTask` level.
+
+In the following example, `pipelineTask` add declares `results` and its default value. Now, `x` is an `optional` param
+in `multiply` task and referring to a task result of `add`. If `add` fails for some reason, finally task `multiply`
+continues execution with default `sum` which is `0`:
+
+```yaml
+spec:
+  tasks:
+    - name: add
+      taskRef:
+        name: add-task
+      params:
+        - name: first
+          value: $(params.first)
+        - name: second
+          value: $(params.second)
+      results:
+        - name: sum
+          value: 0
+  finally:
+    - name: multiply
+      taskRef:
+        name: multiply
+      params:
+        - name: x
+          value: $(task.add.results.sum)
+```
+
+No default for a `required` task param (referring to a task result) allowed. If a task result is missing, `pipeline`
```

Comment:
I'm afraid I don't understand the reason behind this restriction.

It should be up to the pipeline author whether to specify the default or not.
Assuming the pipeline about, you could have the following combinations:

x param type | result default | multiply execution
--------------|--------------|-------------------
mandatory | supplied | executed
mandatory | not supplied | skipped
optional | supplied | executed
optional | not supplied | executed